### DISCUSSION TITLE
ArgoCD Application Controller - Refresh Throttling

### DISCUSSION THREAD

**Discussion by Aenima4six2** [argoproj/argo-cd#6605] [Q&A]
Hello there,
   Recently my team and I have noticed an increase in memory and cpu usage coming from the ArgoCD Application Controller. From what I can gather, there seems to be excessive controller "refreshes" and "reconciliation" occurring due to  resources in the cluster changing in real time. Often times, resource changes are from KEDA `ScaledObjects` (which create HPAs), `ExternalSecret` updates and `Stakater/Reloader` deployment reload updates. From observing the logs and scanning through the source, this behavior is what drives the realtime "sync" status of the app, however, many times the controller performs back to back refreshes within a few milliseconds, causing very high resource usage (over `1000m` cpu and up to `4GB` of memory).  My question is, is it possible to throttle or rate limit processing these events? I am open to any other work arounds or suggestions, as this problem will likely get worse for us, as we are in the process of converting legacy applications to OpenShift/K8s, and have a many more apps that will be added. Also, I am happy to create an issue if what I have described is not expected/desired behavior.

Other note worthy points of interest: 

1. For any given client facing application, we deploy out to three discrete clusters (`dev`, `staging` and `prod`), with ArgoCD tracking an `Application` for each cluster/environment. 
2. We are following the `app-of-apps` pattern with single ArgoCD dedicated repo. The repo hosts a Helm chart, with a discrete `Application` for each of the aforementioned clusters/environments. 
2. We are using OpenShift to host our customer facing applications and OKD for internal applications. ArgoCD is hosted on OKD using the RedHat GitOps operator.

**Comment by wtam2018** [comment]
I can only think of using resource limits https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/.  The next version of GitOps Operator (v1.2) will support resource requests/limits on the argocd workload.

**Reply by Aenima4six2** [reply]
Unfortunately, a memory limit would result in excessive controller restarts since the resource isn't compressible.
